tensorboard/webapp/deeplink/hash_test.ts

```diff
@@ -0,0 +1,94 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {TestBed} from '@angular/core/testing';
+
+import {HashStorageComponent} from './hash';
+
+/** @typehack */ import * as _typeHackStore from '@ngrx/store';
+/** @typehack */ import * as _typeHackStoreTesting from '@ngrx/store/testing';
+
+describe('hash storage test', () => {
+  let store: MockStore<State>;
+  let dispatchSpy: jasmine.Spy;
+  let setStringSpy: jasmine.Spy;
+  let getStringSpy: jasmine.Spy;
+
+  beforeEach(async () => {
+    await TestBed.configureTestingModule({
+      imports: [NoopAnimationsModule, CommonModule],
+      providers: [provideMockStore(), HashStorageContainer],
+      declarations: [HashStorageContainer, HashStorageComponent],
+    }).compileComponents();
+    store = TestBed.get(Store);
+    dispatchSpy = spyOn(store, 'dispatch');
+
+    setStringSpy = jasmine.createSpy();
+    getStringSpy = jasmine.createSpy();
+
+    // Cannot safely stub out window.location.hash or rely on test framework
+    // to not make use of the hash (it does).
+
+    // Do not rely on Polymer bundle in the test.
+    const createElement = spyOn(document, 'createElement');
+    createElement.withArgs('tf-storage').and.returnValue({
+      tf_storage: {
+        setString: setStringSpy,
+        getString: getStringSpy,
+      },
+    } as any);
+    createElement.withArgs('tf-globals').and.returnValue({
+      tf_globals: {
+        setUseHash: jasmine.createSpy(),
+      },
+    } as any);
+    createElement.and.callThrough();
+  });
+
+  it('sets the hash to plugin id by replacing on first load', () => {
+    store.overrideSelector(getActivePlugin, 'foo');
+    const fixture = TestBed.createComponent(HashStorageContainer);
+    fixture.detectChanges();
+
+    expect(setStringSpy).toHaveBeenCalledWith(jasmine.any(String), 'foo', {
```

Comment:
Is there value in asserting that the key argument used is `'__tab__'` by using a `TEST_ONLY`? Same in the following parts in this file.